scipy/optimize/optimize.py

```diff
@@ -984,29 +1065,43 @@ def _minimize_bfgs(fun, x0, args=(), jac=None, callback=None,
     norm : float
         Order of norm (Inf is max, -Inf is min).
     eps : float or ndarray
-        If `jac` is approximated, use this value for the step size.
+        If `jac is None` the absolute step size used for numerical
+        approximation of the jacobian via forward differences.
+    finite_diff_rel_step : None or array_like, optional
+        If `jac in ['2-point', '3-point', 'cs']` the relative step size to
+        use for numerical approximation of the jacobian. The absolute step
+        size is computed as ``h = rel_step * sign(x0) * max(1, abs(x0))``,
+        possibly adjusted to fit into the bounds. For ``method='3-point'``
+        the sign of `h` is ignored. If None (default) then step is selected
+        automatically.
 
     """
     _check_unknown_options(unknown_options)
-    f = fun
-    fprime = jac
-    epsilon = eps
     retall = return_all
 
     x0 = asarray(x0).flatten()
     if x0.ndim == 0:
         x0.shape = (1,)
     if maxiter is None:
         maxiter = len(x0) * 200
-    func_calls, f = wrap_function(f, args)
 
-    old_fval = f(x0)
+    sf = _prepare_scalar_function(fun, x0, jac, args=args, epsilon=eps,
+                                  finite_diff_rel_step=finite_diff_rel_step)
 
-    if fprime is None:
-        grad_calls, myfprime = wrap_function(approx_fprime, (f, epsilon))
-    else:
-        grad_calls, myfprime = wrap_function(fprime, args)
+    f = sf.fun
+    myfprime = sf.grad
+
+    old_fval = f(x0)
     gfk = myfprime(x0)
+
+    if not np.isscalar(old_fval):
```

Comment:
Same as above - isscalar is usually the wrong approach.

Reply:
For better or for worse I think the historical behaviour of most minimizers was to work with an objective function `f` that returned any of `1.1`, `np.array(1.1)`, `np.array([1.1])`. What we need to filter out is `np.array([1.1, 2.2])`, etc. The code as it's currently written does what is required.

One could change to:

```
if np.ndim(df):
    try:
        # required to allow np.array([1.1]) to work.
        df = df.item()
    except ValueError:
            raise ValueError("The user-provided "
                             "objective function must "
                             "return a scalar value.") 
```

However, this has it's own problems; If a function doesn't have an explicit return then it's return value is None:

```
>>> np.ndim(None)
0
>>> np.isscalar(None)
False
```

There are several test cases that don't work if the `np.array([1.1])` is returned by the objective and if `df.item()` is removed, 7dd14d567a8248be3956fd09cb822b0dde502a1. `df.item()` converts df to a standard Python scalar. If `np.array([1.1])` is not permitted at this point I think that would cause quite a few issues downstream as back compatibility would be broken.

What are the use cases where the current behaviour causes problems for users?

Reply:
Your commit ref there doesn't seem to have linked, sadly.

Using isscalar will rule out other number types like Decimal and Fraction, which doesn't look like part of the intent of this patch.

There are very few problems for which `isscalar` is an appropriate solution. Will follow up later on the dangers of item

Reply:
Hopefilly this works: https://github.com/scipy/scipy/commit/07dd14d567a8248be3956fd09cb822b0dde502a1#diff-c1c2c73ad9e5ddef5673e2e95d19020f

Reply:
Would coercion to a float, `df = float(df)`, be acceptable?

Reply:
Ideally I'd go for:
```
f0 = asarray(f0)
if f0.ndim != 0:
    try:
        f0 = f0.squeeze(axis=range(f0.ndim))
    except ValueError:
        raise ValueError("The user-provided ...")
    warnings.warn(UserWarning, "User-provided function returned an array, which was cast to a scalar", stacklevel=2)
```

Reply:
Ah, I now see this is just moved code.

Why does the check not live in `approx_derivative`?

Reply:
That code is fairly understandable, but I can't see why a simple float cast wouldn't work

Reply:
In BFGS the user can provide a jacobian, so a test for scalar-ness is not necessarily done in `approx_derivative`, or in `_differentiable_functions.ScalarFunction`

Reply:
@andyfaff: a float cast would downcast `longdouble`, in case users want more precision

Reply:
> In BFGS the user can provide a jacobian, so approx_derivative is not necessarily called.

Apologies, that comment was intended for the other instance of this in `approx_fprime`

Reply:
> a float cast would downcast longdouble and Decimal

The stumbling block there is the [numerical differentiation code is designed to work in float](https://github.com/scipy/scipy/blob/master/scipy/optimize/_numdiff.py#L12).

Reply:
`approx_fprime` isn't really used for anything after this PR. It's just the public API for a simple scalar function differentiation.
The check is retained here because `approx_derivative` can differentiate functions returning non-scalar arrays, but we want to limit `approx_fprime` to the scalar case only.

Reply:
If nothing else, it seems sloppy to convert back and forth between numpy scalars and python scalars